include/tscore/ver_ptr.h

```diff
@@ -0,0 +1,364 @@
+          //
+          std::memset(reinterpret_cast<char *>(&_vp) + sizeof(_vp), 0, Pad);
+        }
+      }
+
+      explicit Type_(Atomic_access_type const &aa) { _aa = aa; }
+
+      void *
+      ptr() const
+      {
+        return _vp.ptr;
+      }
+
+      void
+      ptr(void *p)
+      {
+        _vp.ptr = p;
+      }
+
+      unsigned
+      version() const
+      {
+        return _vp.version;
+      }
+
+      void
+      version(unsigned v)
+      {
+        _vp.version = v;
+      }
+
+    private:
+      union {
+        VPS _vp;
+        AAT _aa;
+      };
+
+      static_assert(sizeof(_vp) <= sizeof(_aa));
+    };
+
+    namespace
+    {
+      unsigned const Version_bits  = 15;
+      unsigned const Version_shift = 48;
+      constexpr unsigned
+      Mask_version(unsigned v)
+      {
+        return v & ((1U << (Version_bits + 1)) - 1);
+      }
+      std::uint64_t const Version_mask = static_cast<std::uint64_t>(Mask_version(~0U)) << Version_shift;
+
+    } // end anonymous namespace
+
+    // In this specialization, the version is kept in 15 unused bits in a 64-bit pointer.
+    //
+    template <> class Type_<std::uint64_t, true>
+    {
+    public:
+      using Atomic_access_type = std::uint64_t;
+
+      Type_(){};
+
+      explicit Type_(void *p, unsigned v = 0)
+      {
+        _ptr = reinterpret_cast<std::uint64_t>(p) & ~Version_mask;
+        _ptr |= static_cast<uint64_t>(Mask_version(v)) << Version_shift;
+      }
+
+      explicit Type_(Atomic_access_type const &ptr) { _ptr = ptr; }
+
+      void *
+      ptr() const
+      {
+        return reinterpret_cast<void *>(_ptr & ~Version_mask);
+      }
+
+      void
+      ptr(void *p)
+      {
+        _ptr &= Version_mask;
+        _ptr |= reinterpret_cast<std::uint64_t>(p) & ~Version_mask;
+      }
+
+      unsigned
+      version() const
+      {
+        return Mask_version(static_cast<unsigned>(_ptr >> Version_shift));
+      }
+
+      void
+      version(unsigned v)
+      {
+        _ptr &= ~Version_mask;
+        _ptr |= static_cast<std::uint64_t>(Mask_version(v)) << Version_shift;
+      }
+
+    private:
+      std::uint64_t _ptr;
+    };
+
+    enum Impl_type {
+      DEFAULT,        // Atomic access may not be lock free.
+      VERSION_IN_PTR, // Use bits inside of 64-bit pointer to store version.
+      BITS64,         // VPS structure is less than 64 bits, and 64-bit access is lock free.
+      BITS128         // VPS structure is less than 128 bits, and 128-bit compare-and-exchange is available.
+    };
+
+    constexpr Impl_type
+    Impl()
+    {
+      if (std::atomic<std::uint64_t>::is_always_lock_free && (sizeof(VPS) <= 8)) {
+        return BITS64;
+
+      } else if (CAE128 && (sizeof(VPS) <= 16)) {
+        return BITS128;
+
+      } else if (std::atomic<std::uint64_t>::is_always_lock_free && Ptr64bits15unused && (sizeof(void *) == 8)) {
+        return VERSION_IN_PTR;
+      }
+      return DEFAULT;
+    }
```

Comment:
If I force DEFAULT, I get these link error messages:
```
/opt/rh/devtoolset-8/root/usr/include/c++/8/atomic:250: undefined reference to `__atomic_load_16'
/Users/wkaras/REPOS/TS2/src/tscore/.libs/libtscore.so: undefined reference to `__atomic_compare_exchange_16'
```
Also, I don't have a build setup on a 32-bit architecture, so I can't test BITS64.  But, it looks like ATS won't currently build for a 64-bit system with no CAS:  https://github.com/apache/trafficserver/blob/95d1069a9c5f70fe8b0fb57feec95e559ab92b91/include/tscore/ink_queue.h#L167 .  For 32-bit architectures, ATS seems to ignore the need for atomic access:  https://github.com/apache/trafficserver/blob/95d1069a9c5f70fe8b0fb57feec95e559ab92b91/include/tscore/ink_queue.h#L135 .  I don't see how that could work.